text/trace/0234-payload-collection.md

```diff
@@ -0,0 +1,363 @@
+  supported encodings, in an extendable fashion.
+- Aggregating multiple data chunks (to assist with handling asynchronous buffers)
+
+The APIs given here are only a draft to describe general characteristics.
+Like in the semantics, a further discussion will be required before creating actual specifications.
+
+#### `Payload` class
+
+We propose adding a class that will provide the related functionality for handling payload data.
+Both traces and logs could use instances of this class, which will assist sharing related functionality.
+
+For simplicity, we will show a basic example of this class in Python. In reality, an API and its implementation should be separated.
+
+```python
+class Payload:
+    def __init__(self, content: Any, encoding=None, size=None):
+        self.content = self._parse_payload_content(content)
+        self.encoding = encoding
+        self.size = size
+
+    @staticmethod
+    def _parse_payload_content(content: Any) -> Any:
+        # Potentially altering of the content could be added, such as
+        #  format conversion, shortening, filtering, etc.
+        return content
+
+    @classmethod
+    def from_binary(
+        cls,
+        payload_buffer: bytes,
+        encoding: str = 'utf-8',
+    ) -> Payload:
+        if encoding == 'utf-8':
+            content = payload_buffer.decode('utf-8')
+        elif encoding == 'utf-16':
+            content = payload_buffer.decode('utf-16')
+        elif encoding == 'json':
+            content = json.loads(payload_buffer.decode('utf-8'))
+        # ... other supported encodings
+        else:
+            raise ValueError("Unsupported payload encoding %s", encoding)
+
+        return cls(content, encoding=encoding, size=len(payload_buffer))
+```
+
+### Payload attributes in Traces
+
+In order to efficiently collect payload data in traces, this proposal depends on the [support of maps and heterogeneous arrays in spans
+attributes](https://github.com/open-telemetry/opentelemetry-specification/pull/2888).
+
+For adding payload attributes, we propose adding a dedicated method for `Span` class:
+
+```python
+
+def add_payload(self, key: str, payload: Payload):
+  self.set_attributes(
+    {
+      f"{key}.content": payload.content,
+      f"{key}.encoding": payload.encoding,
+      f"{key}.size": payload.size
+    }
+  )
+```
+
+### Payload attributes in Logs
+
+The payload content could be technically collected both as the the log record body, or as an attribute.
+We think that it's needed to specify the preferred way following a discussion in the logs SIG.
+[This](https://github.com/open-telemetry/opentelemetry-specification/pull/2926) PR references this issue as well.
+
+### Null value
+
+Representing Null values in payload contents is required, as it part of JSON and many other formats
+like Protobuf and Avro. Therefore we propose supporting it by spans and logs attributes APIs (while
+in OTel proto schema it is already supported).
```

Comment:
Where is this supported in OTel proto schema?  Do you mean the "AnyValue" being allowed to be "emtpy"?

I'd like more details about where/how you need `Null`  support here.  Are you trying to record payload attributes even if the payload is null?  Could you just write an empty byte array for the Proto/Avro case?

Reply:
The point here is that we need to support JSON payloads like `'{"x": null}'` and also `null`, and similar examples in other formats.
If the payload is encoded with the "AnyValue" type, it can be achieved by using empty objects. Though it's still needed to be supported by the APIs.
In case we add a dedicated proto object, like the `Value` suggested in the alternative, we should also make sure it supports Null values.